lib/hsa/mcwamp_hsa.cpp

```diff
@@ -2059,13 +2059,25 @@ class HSADevice final : public KalmarDevice
         std::string str(fun);
         HSAKernel *kernel = programs[str];
         if (!kernel) {
-            size_t kernel_size = (size_t)((void *)size);
-            char *kernel_source = (char*)malloc(kernel_size+1);
-            memcpy(kernel_source, source, kernel_size);
-            kernel_source[kernel_size] = '\0';
-            //std::cerr << "HSADevice::CreateKernel(): Creating kernel: " << fun << "\n";
-            kernel = CreateOfflineFinalizedKernelImpl(kernel_source, kernel_size, fun);
-            free(kernel_source);
+            if (executables.size() != 0) {
+                for (auto executable_iterator : executables) {
+                    HSAExecutable *executable = executable_iterator.second;
+
+                    // Get symbol handle.
+                    hsa_status_t status;
+                    hsa_executable_symbol_t kernelSymbol;
+                    status = hsa_executable_get_symbol(executable->hsaExecutable, NULL, fun, agent, 0, &kernelSymbol);
+                    STATUS_CHECK(status, __LINE__);
```

Comment:
logic here doesn't seem correct to me. in case there are multiple HSAExecutable instances associated with one HSADevice, there might be cases where `hsa_executable_get_symbol()` may return error, but the for loop shall continue to find the the symbol for the next HSAExecutable until all HSAExecutable has been traversed.